src/consensus/progress_map.zig

```diff
@@ -0,0 +1,1169 @@
+const std = @import("std");
+const sig = @import("../sig.zig");
+
+const Slot = sig.core.Slot;
+const Hash = sig.core.Hash;
+const Pubkey = sig.core.Pubkey;
+const PubkeyArraySet = std.AutoArrayHashMapUnmanaged(Pubkey, void);
+
+/// TODO: any uses of these types are to be evaluated in their context, and
+/// the actual required synchronization semantics are to be determined later.
+const stubs = struct {
+    fn Arc(comptime T: type) type {
+        return struct { arc_ed: T };
+    }
+    fn RwLock(comptime T: type) type {
+        return struct { rwlock_ed: T };
+    }
+};
+
+// TODO: move this somewhere better?
+/// NOTE: usage of this type indicates that the integer
+/// must only be operated on using saturating arithmetic.
+/// This is used to annotate types as saturating, to match
+/// agave's type definitions.
+fn Saturating(comptime T: type) type {
+    return enum(T) {
+        _,
+
+        pub fn toInt(self: Saturating(T)) T {
+            return @intFromEnum(self);
+        }
+
+        /// Returns `self`, casted to a pointer to `T`, retaining all attributes.
+        pub fn asInt(self: anytype) AsInt(@TypeOf(self)) {
+            return @ptrCast(self);
+        }
+
+        pub fn AsInt(comptime Self: type) type {
+            const lazy = struct {
+                fn nonMatchErr() noreturn {
+                    @compileError("Expected single item pointer to " ++ @typeName(Saturating(T)));
+                }
+            };
+
+            var info = switch (@typeInfo(Self)) {
+                .Pointer => |info| info,
+                .Optional => |info| switch (@typeInfo(info.child)) {
+                    .Pointer => return ?AsInt(info.child),
+                    else => lazy.nonMatchErr(),
+                },
+                else => lazy.nonMatchErr(),
+            };
+            if (info.size != .One) lazy.nonMatchErr();
+            if (info.child != Saturating(T)) lazy.nonMatchErr();
+            info.child = T;
+
+            return @Type(.{ .Pointer = info });
+        }
+    };
+}
+
+/// AUDIT: https://github.com/anza-xyz/agave/blob/cb32984a9b0d5c2c6f7775bed39b66d3a22e3c46/core/src/consensus/progress_map.rs
+pub const ProgressMap = struct {
```

Comment:
Could you add some of the logic from agave's ProgressMap? You can add them all, or use your judgement about whether some of them are redundant/pointless, but at least `get_leader_propagation_slot_must_exist` and `get_bank_prev_leader_slot` seem worthwhile.

Reply:
I think it would be better to do this in a follow-up PR, the size of this PR is reaching a boiling point.